python/ray/worker.py

```diff
@@ -1988,6 +1988,48 @@ def export_remote_function(function_id, func_name, func, num_return_vals,
   worker.redis_client.rpush("Exports", key)
 
 
+def in_ipython():
+  """Return true if we are in an IPython interpreter and false otherwise."""
+  try:
+    __IPYTHON__
+    return True
+  except NameError:
+    return False
+
+
+def compute_function_id(func_name, func):
+  """Compute an function ID for a function.
+
+  Args:
+    func_name: The name of the function (this includes the module name plus the
+      function name).
+    func: The actual function.
+
+  Returns:
+    This returns the function ID.
+  """
+  function_id_hash = hashlib.sha1()
+  # Include the function name in the hash.
+  function_id_hash.update(func_name.encode("ascii"))
+  try:
+    # Include the source code in the hash. This will fail if we are in a
+    # regular Python interpreter (but not IPython) because in that case
+    # the source code is not accessible.
+    function_id_hash.update(inspect.getsource(func).encode("ascii"))
+  except (OSError, IOError):
+    # This means that we are in a regular Python interpreter. Verify that
+    # this is actually the case.
+    import __main__ as main
```

Comment:
If it's not too expensive to run `in_ipython` multiple times, maybe we can change this from a `try...except` to just `if not in_ipython()`? Also, I would move the two lines above the `in_ipython` assertion into the `in_ipython` function.